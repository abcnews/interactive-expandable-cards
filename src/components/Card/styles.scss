$transitionShadow: box-shadow 0.25s cubic-bezier(0.25, 0.8, 0.25, 1);
$radius: 0.125rem;

// We want to use flexbox here so that the auto vertical margin on titles
// effectively centers them in rows of cards with variable #s of lines,
// however, a bug in IE's flexbox implementation  means that the cards'
// height is stretched to the natural height of its thumbnail image.
// To work around this, we only declare display: flex on all other
// browsers, using selector hacks (see below).

.card {
  overflow: visible;
  display: flex;
  flex: 1 0 auto;
  flex-direction: column;
  justify-content: flex-start;
  align-items: stretch;
  position: relative;
  z-index: 1;
  margin: 0 0 2rem;
  border: none;
  border-radius: $radius;
  padding: 0.75rem 0.75rem 1.75rem;
  background-color: var(--expandable-cards-background-color, #fff);
  box-shadow: var(--expandable-cards-card-shadow);
  color: var(--expandable-cards-color, #333);
  font-family: ABCSans, sans-serif;
  text-align: center;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
  transition: $transitionShadow;
  min-height: 1px; // IE bugfix

  &:focus {
    outline: none;
    box-shadow: var(--expandable-cards-card-shadow-active);
  }

  &.open {
    border-radius: $radius $radius 0 0;
  }
}

.details {
  width: 100%;
}

.label {
  position: relative;
  margin: -0.75rem -0.75rem 0 -0.75rem;
  padding: 0.3125rem 0.5rem;
  border-radius: $radius $radius 0 0;
  background-color: var(--card-heading-bg, #000);
  color: var(--card-heading-text, #fff);
  font-size: 1rem;
  line-height: 1.1875rem;
  text-transform: uppercase;
  text-shadow: 0 0 1px rgba(0, 0, 0, 0.5);
}

.long {
  font-size: 0.6875rem;
}

.title {
  // !important === ABC News CSS override
  margin: auto -0.25rem !important;
  padding-top: 0.5rem;
  font-size: 1.125rem !important;
  font-weight: normal !important;
  line-height: 1.3;
  min-height: 3.5em;
  height: auto;
  text-transform: uppercase;
  display: flex;
  flex-direction: column;
  justify-content: center;

  &:first-child {
    margin-top: 0 !important;

    .siblingsHaveLabels > & {
      margin-top: 1.75rem !important;
    }
  }

  & > span {
    &::before {
      content: '(';
    }

    &::after {
      content: ')';
    }
  }
}

figure {
  &:first-child {
    margin-top: 0 !important;

    .siblingsHaveLabels > & {
      margin-top: 1.75rem !important;
    }
  }
}

.toggle {
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 0 0 $radius $radius;
  border-top-width: 0;
  border-bottom: 2px solid transparent;
  transition: background-color 0.125s, bottom 0.25s, $transitionShadow;

  &::before,
  &::after {
    content: '';
    display: block;
    position: absolute;
    bottom: -0.75rem;
    left: 50%;
    border-radius: 50%;
    margin-left: -1rem;
    width: 2rem;
    height: 2rem;
  }

  &::before {
    z-index: -1;
    box-shadow: var(--expandable-cards-tab-shadow);

    .root:focus & {
      box-shadow: var(--expandable-cards-tab-shadow-active);
    }
  }

  &::after {
    content: '╲╱';
    font-size: 8px;
    line-height: 4; /* IE can't do rem unit line-height on pseudo elements */
    background-color: var(--card-heading-bg, #000);
    color: var(--card-heading-text, #fff);
    text-shadow: 0 -1px 0 var(--card-heading-text, #fff), 0 -0.5px 0 var(--card-heading-text, #fff),
      0 0.5px 0 var(--card-heading-text, #fff), 0 1px 0 var(--card-heading-text, #fff);
  }

  .open & {
    background-color: var(--card-heading-bg, #000);

    &::before {
      box-shadow: var(--expandable-cards-tab-shadow);
    }

    &::after {
      content: '╱╲';
      transition: background-color 0.25s, background-image 0.25s;
    }
  }

  .open:focus & {

    &::before {
      box-shadow: var(--expandable-cards-tab-shadow-active);
    }
  }
}

.of1 {
  width: 100%;
}

.of2 {
  width: 50%;
}

.of3 {
  width: 33.33%;
}

.of4 {
  width: 25%;
}

.of5 {
  width: 20%;
}

.of6 {
  width: 16.66%;
}
