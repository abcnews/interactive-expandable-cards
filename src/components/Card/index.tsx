import { h, FunctionalComponent, Fragment } from 'preact';
import classNames from 'classnames';
import { Detail } from '../Detail';
import styles from './styles.scss';
import { blackOrWhiteText, hexToRGB, slug } from '../../lib/utils';
import { useEffect, useRef } from 'preact/hooks';
import { TITLE_SCROLL_MARGIN } from '../../lib/constants';
import CardImage from '../CardImage';
import { ExpandableCardsImage } from '../ExpandableCards';

type CardProps = {
  instance: number;
  index: number;
  title: string;
  image: ExpandableCardsImage | null;
  label: string | null;
  detail: HTMLElement[];
  colour: string;
  shouldTintPhoto: boolean;
  isOpen: boolean;
  siblingsHaveLabels: boolean;
  onNavigate: (event: KeyboardEvent) => void;
  onToggle: () => void;
  itemsPerRow: number;
};

const Card: FunctionalComponent<CardProps> = ({
  instance,
  index,
  title,
  image,
  label,
  detail,
  colour: cardColour,
  shouldTintPhoto,
  isOpen,
  siblingsHaveLabels,
  onNavigate,
  onToggle,
  itemsPerRow
}) => {
  const controlId = `ExpandableCards_${instance}__Control_${index}`;
  const regionId = `ExpandableCards_${instance}__Region_${index}`;
  const order = 1 + (index % itemsPerRow) + Math.floor(index / itemsPerRow) * itemsPerRow * 2;

  const toggleRef = useRef<HTMLDivElement>(null);

  // Scroll newly opened cards into view.
  useEffect(() => {
    if (isOpen) {
      setTimeout(() => {
        const rect = toggleRef.current?.getBoundingClientRect();
        if (!rect) return;
        const { top, bottom } = rect;
        const windowHeight = window.innerHeight;
        if (top < TITLE_SCROLL_MARGIN || bottom > windowHeight - TITLE_SCROLL_MARGIN) {
          toggleRef.current?.scrollIntoView({ behavior: 'smooth', block: 'center', inline: 'end' });
        }
      }, 250);
    }
  }, [isOpen]);

  const matches = title.match(/(.*)\((.*)\)(.*)?/); // Does the title contain braces?
  const titleChildren = matches ? [matches[1], <span>{matches[2]}</span>, matches[3] || ''] : title;
  const cardColourRGB = hexToRGB(cardColour);
  const textColour = blackOrWhiteText(cardColourRGB);

  return (
    <Fragment>
      <dt role="heading" aria-level={3} className={styles[`of${itemsPerRow}`]} style={{ order: order }}>
        <button
          id={controlId}
          aria-controls={regionId}
          aria-expanded={isOpen ? 'true' : 'false'}
          className={classNames(styles.card, {
            [styles.open]: isOpen,
            [styles.siblingsHaveLabels]: siblingsHaveLabels
          })}
          onClick={onToggle}
          onKeyDown={onNavigate}
          style={{ '--card-heading-bg': cardColour, '--card-heading-text': textColour }}
          data-component="Control"
        >
          {label ? (
            <div
              className={classNames(styles.label, { [styles.long]: label.length > 12 })}
              data-expandable-cards-label={slug(label)}
            >
              {label}
            </div>
          ) : null}
          {image && <CardImage image={image} tint={shouldTintPhoto && cardColourRGB} />}
          <div class={styles.title}>{titleChildren}</div>
          <div ref={toggleRef} class={styles.toggle} role="presentation">
            <svg fill="none" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
              <path d="m20 8-8 8-8-8" stroke="currentColor" stroke-width="2" />
            </svg>
          </div>
        </button>
      </dt>
      <dd
        id={regionId}
        role="region"
        aria-hidden={isOpen ? 'false' : 'true'}
        aria-labelledby={controlId}
        className={styles.details}
        style={{ order: order + itemsPerRow, '-webkit-order': order + itemsPerRow }}
      >
        <Detail nodes={detail} open={isOpen} />
      </dd>
    </Fragment>
  );
};

export default Card;
