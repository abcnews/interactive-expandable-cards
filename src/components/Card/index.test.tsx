import { h } from 'preact';
import render from 'preact-render-to-string';

import Card from '.';

describe('Card', () => {
  test('It renders', () => {
    const actual = render(
      <Card
        instance={0}
        index={0}
        title={'Title'}
        label={'Label'}
        image={{
          alt: 'Test',
          url: 'https://example.com/test.jpg',
          renditions: []
        }}
        detail={[document.createElement('p')]}
        siblingsHaveLabels={false}
        colour={'#000'}
        shouldTintPhoto={false}
        itemsPerRow={3}
        isOpen={false}
        onToggle={() => {}}
        onNavigate={ev => {}}
      />
    );
    expect(actual).toMatchSnapshot();
  });
});
