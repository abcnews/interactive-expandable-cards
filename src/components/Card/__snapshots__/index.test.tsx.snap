// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card It renders 1`] = `"<dt role=\\"heading\\" aria-level=\\"3\\" style=\\"order: 1;\\"><button id=\\"ExpandableCards_0__Control_0\\" aria-controls=\\"ExpandableCards_0__Region_0\\" aria-expanded=\\"false\\" class style=\\"--card-heading-bg: #000; --card-heading-text: white;\\" data-component=\\"Control\\"><div class data-expandable-cards-label=\\"label\\">Label</div><figure role=\\"presentation\\"><img alt=\\"Test\\" src=\\"https://example.com/test.jpg\\" srcset sizes=\\"(min-width: 1200px) 180px, (max-width: 1200px; min-width: 992px) 145px, 210px\\" /></figure><div>Title</div><div role=\\"presentation\\"><svg fill=\\"none\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"m20 8-8 8-8-8\\" stroke=\\"currentColor\\" stroke-width=\\"2\\"></path></svg></div></button></dt><dd id=\\"ExpandableCards_0__Region_0\\" role=\\"region\\" aria-hidden=\\"true\\" aria-labelledby=\\"ExpandableCards_0__Control_0\\" style=\\"order: 4; -webkit-order: 4px;\\"><div class data-component=\\"Detail\\"><div class=\\"undefined u-richtext\\"></div></div></dd>"`;
